src/arithmetic/algebraic_expression.c

```diff
@@ -203,41 +203,81 @@ AlgebraicExpressionResult *AlgebraicExpression_Execute(AlgebraicExpression *ae)
-            A = C;
 
-            // Graph matrices mustn't be modified, restore to original form.
-            // TODO: Verify that descriptor indeed changes underline input.
-            // if(term.transpose) GrB_transpose(B, NULL, NULL, B, NULL);
+        AlgebraicExpressionOperand leftTerm;
+        AlgebraicExpressionOperand rightTerm;
+
+        while(ae->operand_count > 1) {
+            // Pick matrix with min NNZ.
+            int minMatPos = _AlgebraicExpression_MinMatPos(ae);
+            int secondOperandPos = -1;
+            // Pick second operand, either left or right of minMatPos.
+            if(minMatPos == ae->operand_count-1) {
+                // Last operand, multiply to the right.
+                secondOperandPos = minMatPos-1;
+                rightTerm = ae->operands[minMatPos];
+                leftTerm = ae->operands[secondOperandPos];
+            } else if(minMatPos == 0) {
+                // First operand, multiply to the left.
+                secondOperandPos = 1;
+                leftTerm = ae->operands[0];
+                rightTerm = ae->operands[secondOperandPos];
+            } else {
+                /* Intermidate operand, prefer second operand
+                * to have the least NNZ between the two options. */
+                GrB_Index leftNvals;
+                GrB_Index rightNvals;
+                GrB_Matrix_nvals(&leftNvals, ae->operands[minMatPos-1].operand);
+                GrB_Matrix_nvals(&rightNvals, ae->operands[minMatPos+1].operand);
+                if(leftNvals < rightNvals) {
+                    // Multiply to the right.
+                    secondOperandPos = minMatPos-1;
+                    rightTerm = ae->operands[minMatPos];
+                    leftTerm = ae->operands[secondOperandPos];
+                } else {
+                    // Multiply to the left.
+                    secondOperandPos = minMatPos+1;
+                    leftTerm = ae->operands[minMatPos];
+                    rightTerm = ae->operands[secondOperandPos];
+                }
+            }
+
+            // Terms must be consecutive.
+            assert(abs(minMatPos - secondOperandPos) == 1);
+
+            // Multiply and reduce.
+            if(leftTerm.transpose) GrB_Descriptor_set(desc, GrB_INP0, GrB_TRAN);
+            if(rightTerm.transpose) GrB_Descriptor_set(desc, GrB_INP1, GrB_TRAN);
+
+            _AlgebraicExpression_Execute_MUL(C, leftTerm.operand, rightTerm.operand, desc);
 
             // Restore descriptor to default.
             GrB_Descriptor_set(desc, GrB_INP0, GxB_DEFAULT);
             GrB_Descriptor_set(desc, GrB_INP1, GxB_DEFAULT);
+
+            // Assign result to minMatPos position and shrink operands array.
+            ae->operands[minMatPos].operand = C;
+            ae->operands[minMatPos].transpose = false;
+
+            // Shift left.
+            for(int i = secondOperandPos; i < ae->operand_count-1; i++)
+                ae->operands[i] = ae->operands[i+1];
```

Comment:
`memmove` is an option here too (though not likely to noticeably affect performance).